mnemonic-collections/src/main/java/org/apache/mnemonic/collections/DurableHashMapImpl.java

```diff
@@ -0,0 +1,447 @@
+  private MemChunkHolder<A> chunkAddr;
+  private A allocator;
+  /**
+   * Set initial capacity for a hashmap. It can grow in size.
+   * 
+   * @param capacity
+   *          Initial capacity to be set
+   */
+  public void setCapacityHint(long capacity) {
+    if (0 == capacity) {
+      totalCapacity = DEFAULT_MAP_SIZE;
+    } else {
+      totalCapacity = 1;
+      while (totalCapacity < capacity) {
+        totalCapacity <<= 1;
+      }
+    }
+    threshold = (long) (totalCapacity * DEFAULT_MAP_LOAD_FACTOR);
+  }
+
+  /**
+   * Add a new key-value pair to map
+   * 
+   * @param key
+   *          the key to be set
+   *
+   * @param value
+   *          the value to be set
+   *
+   * @return previous value with key else return null
+   */
+  @Override
+  public V put(K key, V value) {
+    int hash = hash(key.hashCode());
+    long bucketIndex = getBucketIndex(hash);
+    long bucketAddr = holder.get() + MAX_OBJECT_SIZE * bucketIndex;
+    V retVal = addEntry(key, value, bucketAddr);
+    if (autoResize && (mapSize >= threshold)) {
+      resize(2 * totalCapacity);
+    }
+    return retVal;
+  }
+
+  /**
+   * Add a new key-value pair to map at a given bucket address
+   * 
+   * @param key
+   *          the key to be set
+   *
+   * @param value
+   *          the value to be set
+   *
+   * @param bucketAddr
+   *          the addr of the bucket where key is hashed
+   *
+   * @return previous value with key else return null
+   */
+  public V addEntry(K key, V value, long bucketAddr) {
+    V retValue = null;
+    long handler = unsafe.getAddress(bucketAddr);
+    if (0L == handler) {
+      DurableSinglyLinkedList<MapEntry<K, V>> head = DurableSinglyLinkedListFactory.create(allocator, 
+          listefproxies, listgftypes, false);
+      MapEntry<K, V> entry = MapEntryFactory.create(allocator, factoryProxy, genericField, false);
+      entry.setKey(key, false);
+      entry.setValue(value, false);
+      head.setItem(entry, false);
+      unsafe.putLong(bucketAddr, head.getHandler());
+      mapSize++;
+    } else {
+      DurableSinglyLinkedList<MapEntry<K, V>> head = DurableSinglyLinkedListFactory.restore(allocator,
+          listefproxies, listgftypes, handler, false);
+      DurableSinglyLinkedList<MapEntry<K, V>> prev = head;
+      boolean found = false;
+      while (null != head) {
+        MapEntry<K, V> mapEntry = head.getItem();
+        K entryKey = mapEntry.getKey();
+        if (entryKey == key || entryKey.equals(key)) {
+          retValue = mapEntry.getValue();
+          mapEntry.setValue(value, false);
+          found = true;
+          break;
+        }
+        prev = head;
+        head = head.getNext();
+      }
+      if (true != found) {
+        DurableSinglyLinkedList<MapEntry<K, V>> newNode = DurableSinglyLinkedListFactory.create(allocator, 
+            listefproxies, listgftypes, false);
+        MapEntry<K, V> entry = MapEntryFactory.create(allocator, factoryProxy, genericField, false);
```

Comment:
In this place, a OutofHybridMemory Exception could be thrown out, It might be better to catch this exception and destory newNode manually if the autoreclaim is flagged as false.